support/linting/src/pallet_index.rs

```diff
@@ -0,0 +1,301 @@
+
+impl Lint for RequireExplicitPalletIndex {
+    fn lint(source: &File) -> Result {
+        let mut visitor = ConstructRuntimeVisitor::new(source.to_token_stream());
+        visitor.visit_file(source);
+
+        if !visitor.errors.is_empty() {
+            return Err(visitor.errors);
+        }
+
+        Ok(())
+    }
+}
+
+struct ConstructRuntimeVisitor {
+    original_tokens: String,
+    errors: Vec<syn::Error>,
+}
+
+impl<'ast> syn::visit::Visit<'ast> for ConstructRuntimeVisitor {
+    fn visit_item_macro(&mut self, node: &'ast syn::ItemMacro) {
+        let is_construct_runtime = node
+            .mac
+            .path
+            .segments
+            .last()
+            .map_or(false, |segment| segment.ident == "construct_runtime");
+
+        if is_construct_runtime {
+            let tokens = node.mac.tokens.clone();
+
+            match syn::parse2::<RuntimeDeclaration>(tokens) {
+                Ok(runtime_decl) => match runtime_decl {
+                    RuntimeDeclaration::Explicit(runtime) => {
+                        self.check_pallets_for_index(&runtime.pallets);
+                    }
+                    RuntimeDeclaration::ExplicitExpanded(runtime) => {
+                        self.check_pallets_for_index(&runtime.pallets);
+                    }
+                    RuntimeDeclaration::Implicit(runtime) => {
+                        // Only implicit runtime allows `None` for index
+                        for pallet in runtime.pallets {
+                            if pallet.index.is_none() {
+                                self.errors.push(syn::Error::new(
+                                    pallet.name.span(),
+                                    format!(
+                                        "Pallet `{}` does not have an explicit index in the implicit construct_runtime!",
+                                        pallet.name.to_token_stream()
+                                    ),
+                                ));
+                            }
+                        }
+                    }
+                },
+                Err(e) => self.errors.push(e),
+            }
+        }
+
+        syn::visit::visit_item_macro(self, node);
+    }
+}
+
+impl ConstructRuntimeVisitor {
+    fn new(original_tokens: impl Into<TokenStream2>) -> Self {
+        ConstructRuntimeVisitor {
+            original_tokens: {
+                let mut st = original_tokens.into().to_string();
+                st.retain(|c| !c.is_whitespace());
+                st
+            },
+            errors: Vec::new(),
+        }
+    }
+
+    fn check_pallets_for_index(
+        &mut self,
+        pallets: &[procedural_fork::exports::construct_runtime::parse::Pallet],
+    ) {
+        for pallet in pallets {
+            // Check for explicit index and detect missing indices
+            if !self
+                .original_tokens
+                .contains(format!("={},", pallet.index).as_str())
```

Comment:
Technically speaking, we really just need to look for the `=` sign, since it MUST also have been followed by a number in order for it to compile, but this here also checks whether the pallet index is what we expect so it's fine?

Reply:
problem is things like `=1` vs `=10`

Reply:
what we're verifying here is that the index _isn't_ an auto-calculated one, so we are verifying that it appears somewhere in the code basically